### PR TITLE
\color affects following \right, put array cells in their own groups

### LINKED ISSUE #1844: Colored bracket.
LaTeX renders

$$\ln\left(x\color{red}\right)y$$

as expected. The closing bracket is red. It's the only character that is red.

With KaTeX there is no red bracket. How can I achieve this?

### PR DESCRIPTION
**Pull request by edemaine**
Fix #1844 by giving `leftright` nodes a `rightColor` attribute for how to color the right bracket.  Now `\color` sets the macro `\current@color` in the current environment (in particular, resetting after `\right`), just like `color.sty` does in LaTeX.  This is used to specially pass the current color into any following `\right` and then into the `leftright` parse node.

To confirm LaTeX behavior, here's how it renders the new ColorImplicit test:

image

(I tried `texcmp`, but it doesn't work very well with colored text; suffice to say the letters are also aligned with our test.)

Along the way, I discovered environment leakage from one cell of an array to the next, unlike in LaTeX. This PR also fixes an issue, so `\def`s no longer incorrectly span multiple cells of an array.

### CODE REVIEW

**Comment by edemaine** [comment]
Failing screenshot tests show that there's currently environment leakage from one cell of an array to the next, unlike in LaTeX.  I'll need to fix this before this can be reviewed/merged.

**Comment by edemaine** [comment]
Fixed! (Turned out to be really easy.) So this PR also fixes an issue with `\def`s incorrectly spanning multiple cells of an array.

**Comment by edemaine** [comment]
OK, group isolation wasn't quite as easy as I thought, because of the messyness that is `nextToken`.  I fixed most cases, but the very first character after `\begin{matrix}` is not in a group like it should be. Fixing that will be painful... and this PR is definitely an improvement (currently there is no isolation!), so probably worth merging for now.

File: test/katex-spec.js [hunk 1]
```diff
@@ -3066,6 +3066,21 @@ describe("A macro expander", function() {
             .toParseLike`11\sqrt[2]{2}11`;
     });
 
+    it("array cells generate groups", () => {
+        expect`\def\x{1}\begin{matrix}\x&\def\x{2}\x&\x\end{matrix}`
+            .toParseLike`\begin{matrix}1&2&1\end{matrix}`;
+    });
+
+    // TODO: This doesn't yet work; before the environment gets called,
```
Comment (kevinbarabash) [inline test/katex-spec.js]: Could open an issue for this and include a link to the issue from this comment?

File: src/functions/color.js [hunk 2]
```diff
@@ -102,7 +102,13 @@ defineFunction({
     handler({parser, breakOnTokenText}, args) {
         const color = assertNodeType(args[0], "color-token").color;
 
-        // If we see a styling function, parse out the implicit body
+        // Set macro \current@color in current namespace to store the current
+        // color, mimicking the behavior of color.sty.
+        // This is currently used just to correctly color a \right
+        // that follows a \color command.
+        parser.gullet.macros.set("\\current@color", color);
```
Comment (kevinbarabash) [inline src/functions/color.js]: macros defined within a group get popped as we exit the group right?

File: src/functions/delimsizing.js [hunk 3]
```diff
@@ -145,10 +145,16 @@ defineFunction({
         // \left case below triggers parsing of \right in
         //   `const right = parser.parseFunction();`
         // uses this return value.
+        const color = context.parser.gullet.macros.get("\\current@color");
+        if (color && typeof color !== "string") {
```
Comment (kevinbarabash) [inline src/functions/delimsizing.js]: Do we give `\current@color` a default value somewhere?

File: test/katex-spec.js [hunk 4]
```diff
@@ -3066,6 +3066,21 @@ describe("A macro expander", function() {
             .toParseLike`11\sqrt[2]{2}11`;
     });
 
+    it("array cells generate groups", () => {
+        expect`\def\x{1}\begin{matrix}\x&\def\x{2}\x&\x\end{matrix}`
+            .toParseLike`\begin{matrix}1&2&1\end{matrix}`;
```
Comment (kevinbarabash) [inline test/katex-spec.js]: Interesting... so before this change, the output would've been `\begin{matrix}1&2&2\end{matrix}`?

**Comment by JMSS-Unknown** [comment]
May I ask what is holding up this pull request? I need to be able to colour one side of a bracket without knowing the size of the contents, so I can't use `\big`, etc...

File: src/functions/delimsizing.js (hunk 3 above)
Comment (edemaine) [reply src/functions/delimsizing.js]: No, the default value is `undefined`, which implicitly translates to `color: inherit` in CSS (i.e., don't change the color, use whatever CSS sets it to).  Note that `rightColor` has a type of `?string`.  I'll add a comment about the meaning of `undefined`.

File: test/katex-spec.js (hunk 4 above)
Comment (edemaine) [reply test/katex-spec.js]: Yes, exactly!

image

File: src/functions/color.js (hunk 2 above)
Comment (edemaine) [reply src/functions/color.js]: Correct. I think this is implied by the current comment `Set macro \current@color in current namespace` but feel free to suggest a better wording.

FWIW, we're already using this mechanism to redefine the meaning of `\\` within array environments. It's quite useful!

File: test/katex-spec.js (hunk 1 above)
Comment (edemaine) [reply test/katex-spec.js]: Done.

**Comment by edemaine** [comment]
Sorry for the delay; I hadn't realized this was blocked on me. I just implemented @kevinbarabash's comments, so it's ready for (re)review.

File: src/functions/delimsizing.js (hunk 3 above)
Comment (kevinbarabash) [reply src/functions/delimsizing.js]: Thanks!

File: test/katex-spec.js (hunk 1 above)
Comment (kevinbarabash) [reply test/katex-spec.js]: Thanks!

**Pull request merged**